# OFYP-Analyse presence and proportion of gene flow between UK insects
## split.sh
Translocation, differences in genome positions etc. can decrease mapping coverage. Hence, the chromosomal copy of the second species is split before aligning it with the reference genome(chromosomal copy of species 1). Split.sh can be run as follows
```
./split.sh species_prefix
```
Ensure relevant raw data files are in the same folder

## make_bedfiles.R
This R script extracts start and end positions of regions with an alignment depth of 2(corresponding to each haplotype of the whole-genome alignment pipeline). Then, it overlaps the regions transitively to merge all the regions together. Finally, it checks which regions are present in the vcf file, and creates bed files based on the output. These bed files are used to generate smaller vcf files(can be easily run using bcftools view), which are required to run [SINGER](https://github.com/popgenmethods/SINGER/tree/main), the bioinformatics tool used to generate Ancestral Recombination Graphs(ARGs). 

The command to run this R script is as follows. The depth files should be generated by samtools depth based on the BAM or SAM file.
```
Rscript make_bedfiles.R depth1.txt depth2.txt species.vcf bedfiles_directory
```
Then, use the bedfiles to process individual vcfs
```
count=$(find "Path/to/bedfiles_dir" -type f | wc -l)
for INDEX in $(seq 1 "$count"); do bcftools view -Ov -R /bedfiles_dir/Species${INDEX}.bed Species.vcf.gz > /indiv_vcf_dir/Species${INDEX}.vcf; done
```

You can generate the depth files using the following command
```
samtools depth species1.bam > depth1.txt
```
